src/hotspot/share/opto/compile.cpp

```diff
@@ -3666,6 +3666,33 @@ void Compile::final_graph_reshaping_main_switch(Node* n, Final_Reshape_Counts& f
 
   case Op_LoadVector:
   case Op_StoreVector:
+#ifdef ASSERT
+#ifdef AMD64
+    // Add VerifyAlignment node between adr and load / store.
+    if (AlignVector && VerifyAlignVector) {
+      bool must_verify_alignment = n->is_LoadVector() ? n->as_LoadVector()->must_verify_alignment() :
+                                                        n->as_StoreVector()->must_verify_alignment();
+      if (must_verify_alignment) {
+        jlong memory_size = n->is_LoadVector() ? n->as_LoadVector()->memory_size() :
+                                                 n->as_StoreVector()->memory_size();
+        // The memory access should be aligned to the vector length in bytes.
+        // However, the underlying array is possibly less well aligned, but at least
+        // to ObjectAlignmentInBytes. Hence, even if multiple arrays are accessed in
+        // a loop we can expect at least the following alignment:
+        jlong alignment = MIN2(memory_size, (jlong)ObjectAlignmentInBytes);
+        assert(2 <= alignment && alignment <= 64, "alignment must be in range");
+        assert(is_power_of_2(alignment), "alignment must be power of 2");
```

Comment:
Do we have somewhere check that vector size is always power of 2 or multiply of object alignment (for case when it is bigger than object alignment)?

Reply:
Yes, at least we do those checks in SuperWord. We only ever create power of 2 vector sizes. And `ObjectAlignmentInBytes` is verified to also always be power of 2.
These asserts here are just there to catch any issues if this assumption is ever broken in the future.